R/distribution_formulas.R

```diff
@@ -0,0 +1,192 @@
+#'   to.
+#' @inheritParams new_data_distribution
+#' @param is_random_effect A boolean indicating whether or not the process is
+#'   estimated as a random effect.
+#' @seealso
+#' * [new_data_distribution()]
+#' @export
+new_process_distribution <- function(par,
+                                     module,
+                                     family,
+                                     sd = list(value = 1, estimated = FALSE),
+                                     is_random_effect = FALSE) {
+  families <- c("lognormal", "gaussian")
+  if (family[["family"]] == "normal") {
+    stop("use family = gaussian() instead")
+  }
+  if (!(family[["family"]] %in% families)) {
+    stop("FIMS currently does not offer this distribution for processes.")
+  }
+  if (family[["family"]] == "lognormal") {
+    new_module <- new(TMBDlnormDistribution)
+    new_module$log_logsd <- new(
+      ParameterVector,
+      log(sd$value),
+      length(sd$value)
+    )
+    new_module$log_logsd$set_all_estimable(sd$estimated)
+  }
+
+  if (family[["family"]] == "gaussian") {
+    new_module <- new(TMBDnormDistribution)
+    new_module$log_sd <- new(ParameterVector, log(sd$value), length(sd$value))
+    new_module$log_sd$set_all_estimable(sd$estimated)
+  }
+
+  module$field(par)$set_all_random(is_random_effect)
+
+  n_dim <- length(module$field(par))
+  new_module$x <- new(ParameterVector, n_dim)
+  new_module$expected_values <- new(ParameterVector, n_dim)
+  for (i in 1:n_dim) {
+    new_module$x[i]$value <- 0
+    new_module$expected_values[i]$value <- 0
+  }
+
+  new_module$set_distribution_links(
+    "random_effects",
+    module$field(par)$get_id()
+  )
+
+  return(new_module)
+}
+
+#' Lognormal family and link specification
+#'
+#' @param link A string specifying the link function for the family.
+#' @return
+#' An object of class "family" (which has a concise print method). Though this
+#' particular family has a truncated length compared to other distributions in
+#' [stats::family()].
+#' \item{family}{character: the family name.}
+#' \item{link}{character: the link name.}
+#' \item{linkfun}{function: the link.}
+#' \item{linkinv}{function: the inverse of the link function.}
+#' \item{mu.eta}{
+#'   TODO: document mu.eta
+#'   function: derivative \eqn{TODO}.
+#' }
+#' \item{valideta}{
+#'   logical function. Returns \code{TRUE} if a linear predictor \code{eta} is
+#'   within the domain of \code{linkinv}.
+#' }
+#' \item{name}{A string gibing the name of the link.}
+#' @seealso
+#' * [stats::family()]
+#' * [stats::gaussian()]
+#' * [multinomial()]
+#' @export
+#' @examples
+#' fam <- lognormal()
+#' fam[["family"]]
+#' fam$link
+lognormal <- function(link = "log") {
+  r <- list(family = "lognormal")
+  f <- c(r, list(link = link), stats::make.link(link))
+  class(f) <- "family"
+  return(f)
+}
+
+#' Multinomial family and link specification
+#'
+#' @param link link function association with family
+#' @return An object of class "family"
```

Comment:
The full list that is returned needs to be documented.

Reply:
done